libs/gen/EnumsSourceGenerator.cs

```diff
@@ -0,0 +1,164 @@
+
+        return method.ToString();
+    }
+
+    private static string GenerateGetEnumDescriptionsMethod(string enumName, List<(string Name, object? Value, AttributeSyntax Description)> values)
+    {
+        var method = new StringBuilder();
+        method.AppendLine($"    /// <summary>");
+        method.AppendLine($"    /// Gets the descriptions of the set flags. Assumes the enum is a flags enum.");
+        method.AppendLine($"    /// If no description exists, returns the ToString() value of the input value.");
+        method.AppendLine($"    /// </summary>");
+        method.AppendLine($"    /// <param name=\"value\">Enum value.</param>");
+        method.AppendLine($"    /// <returns>Array of descriptions.</returns>");
+        method.AppendLine($"    public static string[] Get{enumName}Descriptions({enumName} value)");
+        method.AppendLine("    {");
+        foreach (var (name, value, description) in values)
+        {
+            if (!IsPow2(value))
+            {
+                var toString = description?.ArgumentList?.Arguments.FirstOrDefault()?.ToString() ?? $"\"{name}\"";
+                method.AppendLine($"        if (value is {enumName}.{name}) return [{toString}];");
+            }
+        }
+        method.AppendLine("        var setFlags = BitOperations.PopCount((ulong)value);");
+        method.AppendLine("        if (setFlags == 1)");
+        method.AppendLine("        {");
+        method.AppendLine("            return value switch");
+        method.AppendLine("            {");
+        foreach (var (name, value, description) in values)
+        {
+            if (description is null) continue;
+            if (!IsPow2(value)) continue;
+            method.AppendLine($"                {enumName}.{name} => [{description?.ArgumentList?.Arguments.FirstOrDefault()?.ToString()}],");
+        }
+        method.AppendLine($"                _ => [value.ToString()],");
+        method.AppendLine("            };");
+        method.AppendLine("        }");
+        method.AppendLine();
+        method.AppendLine("        var descriptions = new string[setFlags];");
+        method.AppendLine("        var index = 0;");
+        foreach (var (name, value, description) in values)
+        {
+            if (description is null) continue;
+            if (!IsPow2(value)) continue;
+            method.AppendLine($"        if ((value & {enumName}.{name}) != 0) descriptions[index++] = {description.ArgumentList?.Arguments.FirstOrDefault()?.ToString()};");
+        }
+        method.AppendLine();
+        method.AppendLine("        return descriptions;");
+        method.AppendLine("    }");
+
+        return method.ToString();
+    }
+
+    static bool IsPow2(object? value)
+    {
+        if (value is int x) return x != 0 && (x & (x - 1)) == 0;
+        return false;
+    }
+
+    private record struct EnumDetails(string Namespace, string EnumName, List<(string Name, object? Value, AttributeSyntax Description)> Values);
```

Comment:
I fear this value type data model is not cacheable and thus breaks the incremental generation.
- .NET collection types rarely implement structural equality 
  - See more about the topic here https://github.com/dotnet/runtime/issues/77183 and see ["4. Watch out for collection types"](https://andrewlock.net/creating-a-source-generator-part-9-avoiding-performance-pitfalls-in-incremental-generators/#4-watch-out-for-collection-types) in Andrew Lock's article
  - I recommend using [`EquatableArray<T>`](https://github.com/CommunityToolkit/dotnet/blob/main/src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/EquatableArray%7BT%7D.cs) by Sergio Pedri (and add 3rd party notice to the NOTICE.md) for all collections in a source generation pipeline
- The `AttributeSyntax` instance in the pipeline is not equatable and such breaks the caching too.

Reply:
I am not sure I follow, whether there is a viable fix to each of these issues raised here?

Reply:
The fact that this seems to work already is _weird_. The source generator targeting should be limited by the Visual Studio being stuck in .NET FX but I've seen Visual Studio & Roslyn folks work on making the source generator run out-of-proc so maybe that's already working then? Cool!